fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/LuceneSerializer.java

```diff
@@ -0,0 +1,180 @@
+        return encoded;
+    }
+
+    @Nullable
+    public static byte[] decode(@Nullable byte[] data) {
+        if (data == null) {
+            return null;
+        }
+
+        if (data.length < 2) {
+            throw new RecordCoreException("Invalid data")
+                    .addLogInfo(LogMessageKeys.DATA_VALUE, data);
+        }
+
+        final byte encoding = data[0];
+        final boolean encrypted = (encoding & ENCODING_ENCRYPTED) == ENCODING_ENCRYPTED;
+        final boolean compressed = (encoding & ENCODING_COMPRESSED) == ENCODING_COMPRESSED;
+
+        byte[] decoded = decompressIfNeeded(compressed, data, 1);
+        LOGGER.trace("Decoded data for compressed={}, encrypted={}, encodedDataSize={}, decodedDataSize={}",
+                compressed, encrypted, data.length, decoded.length);
+        return decoded;
+    }
+
+    @Nonnull
+    private static byte[] compressIfNeeded(boolean compressionNeeded, @Nonnull byte[] data, @Nonnull ByteBuffersDataOutput encodedDataOutput,
+                                           @Nonnull EncodingState state, int offset) {
+        if (!compressionNeeded) {
+            return fallBackToUncompressed(encodedDataOutput.toArrayCopy(), data, state, offset);
+        }
+
+        try (Compressor compressor = CompressionMode.HIGH_COMPRESSION.newCompressor()) {
+            encodedDataOutput.writeByte(COMPRESSION_VERSION_FOR_HIGH_COMPRESSION);
+            encodedDataOutput.writeVInt(data.length);
+            compressor.compress(data, 0, data.length, encodedDataOutput);
+            final byte[] compressedData = encodedDataOutput.toArrayCopy();
+
+            // Compress only if it helps to shorten the bytes
+            if (compressedData.length < data.length) {
+                state.setCompressed(true);
+                return compressedData;
+            } else {
+                return fallBackToUncompressed(compressedData, data, state, offset);
+            }
+        } catch (Exception e) {
+            throw new RecordCoreException("Lucene data compression failure", e);
+        }
+    }
+
+    private static byte[] fallBackToUncompressed(@Nonnull byte[] compressedData, @Nonnull byte[] originalData,
+                                                 @Nonnull EncodingState state, int offset) {
+        state.setCompressed(false);
+        final byte[] encoded = new byte[originalData.length + 1];
+        System.arraycopy(compressedData, 0, encoded, 0, offset);
+        System.arraycopy(originalData, 0, encoded, offset, originalData.length);
+        return encoded;
+    }
+
+    @Nonnull
+    private static byte[] decompressIfNeeded(boolean decompressionNeeded, @Nonnull byte[] data, int offset) {
+        if (!decompressionNeeded) {
+            // Return the array to be the final output for decoding, so the offset bytes need to be removed
+            return Arrays.copyOfRange(data, offset, data.length);
+        }
+
+        if (data.length < 2) {
```

Comment:
I guess there should be a `- offset` here, even though `offset` is never very big today.

Reply:
Good catch! Fixed